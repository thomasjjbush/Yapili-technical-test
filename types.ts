import { ReactNode } from 'react';
import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';
import { MockStoreEnhanced } from 'redux-mock-store';

export type Action = ThunkAction<Promise<AnyAction | void>, Store, unknown, AnyAction>;

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface Character {
    comics?: CharecterCollection;
    description?: string;
    error?: boolean;
    events?: CharecterCollection;
    id?: number;
    modified?: string;
    name?: string;
    resourceURI?: string;
    series?: CharecterCollection;
    stories?: CharecterCollection;
    thumbnail?: {
        extension: string;
        path: string;
    };
}

export enum CharacterActions {
    CLEARED = 'character/CLEARED',
    ERROR = 'character/ERROR',
    LOADED = 'character/LOADED',
}

interface CharecterCollection {
    available?: number;
    collectionURI?: string;
    items: Resource[];
    returned?: number;
}

export enum Dimensions {
    LANDSCAPE = 'landscape',
    STANDARD = 'standard',
}

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export interface FallbackImageProps {
    src: string;
    width?: string;
    height?: string;
}

export interface Home extends Partial<MarvelResponse> {
    error?: boolean;
    results?: Character[];
}

export enum HomeActions {
    ERROR = 'home/ERROR',
    LOADED = 'home/LOADED',
    LOADING = 'home/LOADING',
}

export interface ImageProps {
    dimension?: Dimensions;
    extension: string;
    height?: string;
    path: string;
    width?: string;
}

export interface MarvelResponse {
    count: number;
    limit: number;
    offset: number;
    results: any[];
    total: number;
}

export type MockStore = MockStoreEnhanced<Partial<Store>, Dispatch>;

export interface Options {
    [key: string]: string | number;
    apikey?: string;
    limit?: number;
    offset?: number;
}

export interface PaginatedButtonsProps {
    currentPage: number;
    numberOfButtons: number;
    numberOfPages: number;
    skip?: number;
    to: (i: number) => Partial<Location>;
}

export interface Resource {
    resourceURI: string;
    name: string;
    type?: string;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export interface Store {
    character: Character;
    home: Home;
}

export type StyledProps<Props = {}> = { theme: Theme } & Props;

export interface Theme {
    breakpoints: {
        desktop: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        darkGrey: string;
        marvelRed: string;
        mediumGrey: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: HtmlWebPackPlugin[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
