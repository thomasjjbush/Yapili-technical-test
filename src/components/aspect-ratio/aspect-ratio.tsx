import React, { FC, ReactElement } from 'react';
import styled from 'styled-components';
import { AspectRatios, AspectRatioProps as Props, StyledProps } from '../../../types';

const Container = styled.div<StyledProps<{ ratio: AspectRatios }>>`
    padding-bottom: ${({ ratio }) => ratio}%;
    position: relative;
`;

const Child = styled.div`
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
`;

export const AspectRatio: FC<Props> = ({ children, ratio }: Props): ReactElement => (
    <Container ratio={ratio}>
        <Child>{children}</Child>
    </Container>
);
