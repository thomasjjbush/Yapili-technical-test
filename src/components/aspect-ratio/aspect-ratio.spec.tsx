import React from 'react';
import { shallow } from 'enzyme';
import { AspectRatios } from '../../../types';
import { AspectRatio } from './aspect-ratio';

describe('AspectRatio', () => {
    it.each`
        ratio
        ${AspectRatios['16:9']}
        ${AspectRatios['1:1']}
        ${AspectRatios['4:3']}
    `('Should produce correct styles for $ratio', ({ ratio }) => {
        expect(shallow(<AspectRatio ratio={ratio}>Child</AspectRatio>)).toMatchSnapshot();
    });
});
