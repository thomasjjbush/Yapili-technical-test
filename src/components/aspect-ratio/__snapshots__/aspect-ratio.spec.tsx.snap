// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AspectRatio Should produce correct styles for 56.25 1`] = `
<styled.div
  ratio={56.25}
  styles="
    padding-bottom: 56.25%;
    position: relative;
"
>
  <styled.div
    styles="
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
"
  >
    Child
  </styled.div>
</styled.div>
`;

exports[`AspectRatio Should produce correct styles for 75 1`] = `
<styled.div
  ratio={75}
  styles="
    padding-bottom: 75%;
    position: relative;
"
>
  <styled.div
    styles="
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
"
  >
    Child
  </styled.div>
</styled.div>
`;

exports[`AspectRatio Should produce correct styles for 100 1`] = `
<styled.div
  ratio={100}
  styles="
    padding-bottom: 100%;
    position: relative;
"
>
  <styled.div
    styles="
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
"
  >
    Child
  </styled.div>
</styled.div>
`;
